src/test/test-copy.c

```diff
@@ -50,6 +51,123 @@ TEST(copy_file) {
         unlink(fn_copy);
 }
 
+TEST(copy_tree_replace_file) {
+        _cleanup_free_ char *buf = NULL;
+        char *src, *dest;
+        size_t sz;
+        int fd;
+
+        assert_se(tempfn_random("/tmp/test-copy_file.XXXXXX", NULL, &src) >= 0);
+        assert_se(tempfn_random("/tmp/test-copy_file.XXXXXX", NULL, &dest) >= 0);
+
+        assert_se(write_string_file(src, "bar bar", WRITE_STRING_FILE_CREATE) == 0);
+        assert_se(write_string_file(dest, "foo foo foo", WRITE_STRING_FILE_CREATE) == 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "foo foo foo\n"));
+        assert_se(sz == 12);
+
+        /* The file exists- now overwite original contents, and test the COPY_REPLACE flag. */
+
+        assert_se(copy_tree(src, dest, UID_INVALID, GID_INVALID, COPY_REFLINK) < 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "foo foo foo\n"));
+        assert_se(sz == 12);
+
+        assert_se(copy_tree(src, dest, UID_INVALID, GID_INVALID, COPY_REFLINK|COPY_REPLACE) == 0);
+
+        assert_se(read_full_file(dest, &buf, &sz) == 0);
+        assert_se(streq(buf, "bar bar\n"));
+        assert_se(sz == 8);
+}
+
+TEST(copy_tree_replace_dirs) {
+        _cleanup_free_ char *buf = NULL;
+        size_t sz;
+        _cleanup_free_ char *src_path1, *src_path2, *dst_path1, *dst_path2;
+        _cleanup_(rm_rf_physical_and_freep) char *src_directory = NULL;
+        _cleanup_(rm_rf_physical_and_freep) char *dst_directory = NULL;
+        char file1[] = "foo_file";
+        char file2[] = "bar_file";
```

Comment:
I'd inline these two variables

Reply:
I opted not to here because I make use of the string multiple times (i.e. I'd have to inline the raw string in multiple places).

Reply:
Makes sense, but then I think these can be made `const char *` instead of `char[]`